src/game/logic/object/update/laserupdate.cpp

```diff
@@ -0,0 +1,365 @@
+        ParticleSystem *system = g_theParticleSystemManager->Find_Particle_System(m_muzzleParticleSystemID);
+
+        if (system) {
+            system->Set_Position(m_startPos);
+        }
+    }
+
+    if (m_targetParticleSystemID) {
+        ParticleSystem *system = g_theParticleSystemManager->Find_Particle_System(m_targetParticleSystemID);
+
+        if (system) {
+            system->Set_Position(m_endPos);
+        }
+    }
+
+    Coord3D pos;
+
+    if (source_obj) {
+        pos = *source_obj->Get_Position();
+    } else {
+        pos.Set(start_pos);
+        pos.Add(end_pos);
+        pos.Scale(0.5f);
+    }
+
+    Drawable *drawable = Get_Drawable();
+
+    if (drawable) {
+        drawable->Set_Position(&pos);
+    }
+
+    m_dirty = true;
+}
+
+void LaserUpdate::Update_End_Pos()
+{
+    const LaserUpdateModuleData *data = Get_Module_Data();
+    Coord3D oldpos = m_endPos;
+
+    if (m_victimDrawableID) {
+        Drawable *drawable = g_theGameClient->Find_Drawable_By_ID(m_victimDrawableID);
+        bool dead;
+
+        if (drawable && drawable->Get_Object()) {
+            dead = drawable->Get_Object()->Is_Effectively_Dead();
+        } else {
+            dead = false;
+        }
+
+        if (drawable && !dead) {
+            m_endPos = *drawable->Get_Position();
+        } else {
+            if (data->m_punchThroughScalar) {
+                Vector3 pos;
+                pos.Set(m_endPos.x, m_endPos.y, m_endPos.z);
+                pos = pos - Vector3(m_startPos.x, m_startPos.y, m_startPos.z);
+                pos *= data->m_punchThroughScalar;
+                pos = pos + Vector3(m_startPos.x, m_startPos.y, m_startPos.z);
+                m_endPos.x = pos.X;
+                m_endPos.y = pos.Y;
+                m_endPos.z = pos.Z;
+            }
+
+            m_victimDrawableID = DRAWABLE_UNK;
+        }
+
+        if (!(m_startPos == oldpos)) {
+            m_dirty = true;
+        }
+    }
+}
+
+void LaserUpdate::Update_Start_Pos()
+{
+    Coord3D oldpos = m_startPos;
+
+    if (m_sourceDrawableID) {
+        Drawable *drawable = g_theGameClient->Find_Drawable_By_ID(m_sourceDrawableID);
+
+        if (drawable) {
+            if (m_masterBone.Is_Not_Empty()) {
+                Matrix3D m;
+
+                if (!drawable->Get_Current_Worldspace_Client_Bone_Positions(m_masterBone, m)) {
+                    const Coord3D *pos = drawable->Get_Position();
+                    m_startPos.Set(pos);
+                    captainslog_debug("LaserUpdate::updateStartPos() -- Drawable %s is expecting to find a bone %s but "
+                                      "can't. Defaulting to position of drawable.",
+                        drawable->Get_Template()->Get_Name().Str(),
+                        m_masterBone.Str());
+                }
+
+                m_startPos.x = m.Get_X_Translation();
+                m_startPos.y = m.Get_Y_Translation();
+                m_startPos.z = m.Get_Z_Translation();

```

Comment:
Is `Matrix3D` set to something by call to `Get_Current_Worldspace_Client_Bone_Positions` always? I am asking because it is conditioned above and it looks as if `m_startPos` could receive uninitialized values here, which in turn could lead to graphical glitches (there are laser glitches in retail game).

Reply:
This is what the original code does.